docs/DevGuide/Parallelization.md

```diff
@@ -0,0 +1,542 @@
+  What this function does is described below in the discussion of
+  \ref dev_guide_parallelization_phases_of_execution "phases".
+
+There are also several optional members:
+
+- `input_file`: a `static constexpr OptionString` that is the default name of
+  the input file that is to be read. This can be overridden at runtime by
+  passing the `--input-file` argument to the executable.
+- `ignore_unrecognized_command_line_options`: a `static constexpr bool` that
+  defaults to `false`. If set to `true` then unrecognized command line options
+  are ignored. Ignoring unrecognized options is generally only necessary for
+  tests where arguments for the testing framework,
+  [Catch](https://github.com/catchorg/Catch2/), are passed to the executable.
+
+# Phases of an Execution {#dev_guide_parallelization_phases_of_execution}
+
+Global synchronization points, where all cores wait for each other, are
+undesirable for scalability reasons. However, they are sometimes inevitable for
+algorithmic reasons. That is, in order to actually get a correct solution you
+need to have a global synchronization. SpECTRE executables can have multiple
+phases, where after each phase a global synchronization occurs. By global
+synchronization we mean that no parallel components are executing or have more
+tasks to execute: everything is waiting to be told what tasks to perform next.
+
+Every executable must have at least two phases, `%Initialization` and
+`Exit`. The next phase is decided by the static member function
+`determine_next_phase` in the metavariables. Currently this function has access
+to the phase that is
+ending, and also the global cache. In the future we will add support for
+receiving data from various components to allow for more complex decision
+making. Here is an example of a `determine_next_phase` function and the `Phase`
+enum class:
+\snippet Test_AlgorithmCore.cpp determine_next_phase_example
+
+In contrast, an evolution executable might have phases `%Initialization`,
+`SetInitialData`, `Evolve`, and `Exit`, but have a similar `switch` or `if-else`
+logic in the `determine_next_phase` function. The first phase that is entered is
+always `%Initialization`. During the `%Initialization` phase the `initialize`
+function is called on all parallel components. Once all parallel components'
+`initialize` function is complete, the next phase is determined and the
+`execute_next_phase` function is called after on all the parallel components.
+
+At the end of an execution the `Exit` phase has the executable wait to make sure
+no parallel components are performing or need to perform any more tasks, and
+then exits. An example where this approach is important is if we are done
+evolving a system but still need to write data to disk. We do not want to exit
+the simulation until all data has been written to disk, even though we've
+reached the final time of the evolution.
+
+\warning Currently dead-locks are treated as successful termination. In the
```

Comment:
No because the `Exit` phase doesn't call `start_phase`, though this is a good idea! I've added an issue to look at that after this gets merged. At least for the DgElementArray component this check will be easy, I'm not yet sure about how to handle cases like the node writers. Issue #1570